File: Kernel Simulator/Console/Themes/ThemeTools.vb
﻿
'    Kernel Simulator  Copyright (C) 2018-2022  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.IO

Public Module ThemeTools

    ''' <summary>
    ''' All the available built-in themes
    ''' </summary>
    Public ReadOnly Themes As New Dictionary(Of String, ThemeInfo) From {{"Default", New ThemeInfo("_Default")},
                                                                         {"3Y-Diamond", New ThemeInfo("_3Y_Diamond")},
                                                                         {"AyuDark", New ThemeInfo("AyuDark")},
                                                                         {"AyuLight", New ThemeInfo("AyuLight")},
                                                                         {"AyuMirage", New ThemeInfo("AyuMirage")},
                                                                         {"BlackOnWhite", New ThemeInfo("BlackOnWhite")},
                                                                         {"BedOS", New ThemeInfo("BedOS")},
                                                                         {"Bluespire", New ThemeInfo("Bluespire")},
                                                                         {"BrandingBlue", New ThemeInfo("BrandingBlue")},
                                                                         {"BrandingPurple", New ThemeInfo("BrandingPurple")},
                                                                         {"BreezeDark", New ThemeInfo("BreezeDark")},
                                                                         {"Breeze", New ThemeInfo("Breeze")},
                                                                         {"Debian", New ThemeInfo("Debian")},
                                                                         {"GTASA", New ThemeInfo("GTASA")},
                                                                         {"GrayOnYellow", New ThemeInfo("GrayOnYellow")},
                                                                         {"Hacker", New ThemeInfo("Hacker")},
                                                                         {"LinuxColoredDef", New ThemeInfo("LinuxColoredDef")},
                                                                         {"LinuxUncolored", New ThemeInfo("LinuxUncolored")},
                                                                         {"Metallic", New ThemeInfo("Metallic")},
                                                                         {"NeonBreeze", New ThemeInfo("NeonBreeze")},
                                                                         {"NFSHP-Cop", New ThemeInfo("NFSHP_Cop")},
                                                                         {"NFSHP-Racer", New ThemeInfo("NFSHP_Racer")},
                                                                         {"RedConsole", New ThemeInfo("RedConsole")},
                                                                         {"SolarizedDark", New ThemeInfo("SolarizedDark")},
                                                                         {"SolarizedLight", New ThemeInfo("SolarizedLight")},
                                                                         {"TealerOS", New ThemeInfo("TealerOS")},
                                                                         {"TrafficLight", New ThemeInfo("TrafficLight")},
                                                                         {"Ubuntu", New ThemeInfo("Ubuntu")},
                                                                         {"YellowBG", New ThemeInfo("YellowBG")},
                                                                         {"YellowFG", New ThemeInfo("YellowFG")},
                                                                         {"Windows11", New ThemeInfo("Windows11")},
                                                                         {"Windows11Light", New ThemeInfo("Windows11Light")},
                                                                         {"Windows95", New ThemeInfo("Windows95")},
                                                                         {"Wood", New ThemeInfo("Wood")}}

    ''' <summary>
    ''' Sets system colors according to the programmed templates
    ''' </summary>
    ''' <param name="theme">A specified theme</param>
    Public Sub ApplyThemeFromResources(theme As String)
        Wdbg(DebugLevel.I, "Theme: {0}", theme)
        If Themes.ContainsKey(theme) Then
            Wdbg(DebugLevel.I, "Theme found.")

            'Populate theme info
            Dim ThemeInfo As ThemeInfo
            If theme = "Default" Then
                ResetColors()
            ElseIf theme = "NFSHP-Cop" Then
                ThemeInfo = New ThemeInfo("NFSHP_Cop")
            ElseIf theme = "NFSHP-Racer" Then
                ThemeInfo = New ThemeInfo("NFSHP_Racer")
            ElseIf theme = "3Y-Diamond" Then
                ThemeInfo = New ThemeInfo("_3Y_Diamond")
            Else
                ThemeInfo = New ThemeInfo(theme)
            End If

            If Not theme = "Default" Then
#Disable Warning BC42104
                'Set colors as appropriate
                SetColorsTheme(ThemeInfo)
#Enable Warning BC42104
            End If

            'Raise event
            KernelEventManager.RaiseThemeSet(theme)
        Else
            Write(DoTranslation("Invalid color template {0}"), True, ColTypes.Error, theme)
            Wdbg(DebugLevel.E, "Theme not found.")

            'Raise event
            KernelEventManager.RaiseThemeSetError(theme, ThemeSetErrorReasons.NotFound)
        End If
    End Sub

    ''' <summary>
    ''' Sets system colors according to the template file
    ''' </summary>
    ''' <param name="ThemeFile">Theme file</param>
    Public Sub ApplyThemeFromFile(ThemeFile As String)
        Try
            Wdbg(DebugLevel.I, "Theme file name: {0}", ThemeFile)
            ThemeFile = NeutralizePath(ThemeFile, True)
            Wdbg(DebugLevel.I, "Theme file path: {0}", ThemeFile)

            'Populate theme info
            Dim ThemeInfo As New ThemeInfo(New StreamReader(ThemeFile))

            If Not ThemeFile = "Default" Then
                'Set colors as appropriate
                SetColorsTheme(ThemeInfo)
            End If

            'Raise event
            KernelEventManager.RaiseThemeSet(ThemeFile)
        Catch ex As Exception
            Write(DoTranslation("Invalid color template {0}"), True, ColTypes.Error, ThemeFile)
            Wdbg(DebugLevel.E, "Theme not found.")

            'Raise event
            KernelEventManager.RaiseThemeSetError(ThemeFile, ThemeSetErrorReasons.NotFound)
        End Try
    End Sub

    ''' <summary>
    ''' Sets custom colors. It only works if colored shell is enabled.
    ''' </summary>
    ''' <param name="ThemeInfo">Theme information</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    ''' <exception cref="InvalidOperationException"></exception>
    ''' <exception cref="Exceptions.ColorException"></exception>
    Public Function SetColorsTheme(ThemeInfo As ThemeInfo) As Boolean
        If ThemeInfo Is Nothing Then Throw New ArgumentNullException(NameOf(ThemeInfo))

        'Set the colors
        If ColoredShell = True Then
            Try
                InputColor = ThemeInfo.ThemeInputColor
                LicenseColor = ThemeInfo.ThemeLicenseColor
                ContKernelErrorColor = ThemeInfo.ThemeContKernelErrorColor
                UncontKernelErrorColor = ThemeInfo.ThemeUncontKernelErrorColor
                HostNameShellColor = ThemeInfo.ThemeHostNameShellColor
                UserNameShellColor = ThemeInfo.ThemeUserNameShellColor
                BackgroundColor = ThemeInfo.ThemeBackgroundColor
                NeutralTextColor = ThemeInfo.ThemeNeutralTextColor
                ListEntryColor = ThemeInfo.ThemeListEntryColor
                ListValueColor = ThemeInfo.ThemeListValueColor
                StageColor = ThemeInfo.ThemeStageColor
                ErrorColor = ThemeInfo.ThemeErrorColor
                WarningColor = ThemeInfo.ThemeWarningColor
                OptionColor = ThemeInfo.ThemeOptionColor
                BannerColor = ThemeInfo.ThemeBannerColor
                NotificationTitleColor = ThemeInfo.ThemeNotificationTitleColor
                NotificationDescriptionColor = ThemeInfo.ThemeNotificationDescriptionColor
                NotificationProgressColor = ThemeInfo.ThemeNotificationProgressColor
                NotificationFailureColor = ThemeInfo.ThemeNotificationFailureColor
                QuestionColor = ThemeInfo.ThemeQuestionColor
                SuccessColor = ThemeInfo.ThemeSuccessColor
                UserDollarColor = ThemeInfo.ThemeUserDollarColor
                TipColor = ThemeInfo.ThemeTipColor
                SeparatorTextColor = ThemeInfo.ThemeSeparatorTextColor
                SeparatorColor = ThemeInfo.ThemeSeparatorColor
                ListTitleColor = ThemeInfo.ThemeListTitleColor
                DevelopmentWarningColor = ThemeInfo.ThemeDevelopmentWarningColor
                StageTimeColor = ThemeInfo.ThemeStageTimeColor
                ProgressColor = ThemeInfo.ThemeProgressColor
                BackOptionColor = ThemeInfo.ThemeBackOptionColor
                LowPriorityBorderColor = ThemeInfo.ThemeLowPriorityBorderColor
                MediumPriorityBorderColor = ThemeInfo.ThemeMediumPriorityBorderColor
                HighPriorityBorderColor = ThemeInfo.ThemeHighPriorityBorderColor
                TableSeparatorColor = ThemeInfo.ThemeTableSeparatorColor
                TableHeaderColor = ThemeInfo.ThemeTableHeaderColor
                TableValueColor = ThemeInfo.ThemeTableValueColor
                SelectedOptionColor = ThemeInfo.ThemeSelectedOptionColor
                LoadBack()
                MakePermanent()

                'Raise event
                KernelEventManager.RaiseColorSet()
                Return True
            Catch ex As Exception
                WStkTrc(ex)
                KernelEventManager.RaiseColorSetError(ColorSetErrorReasons.InvalidColors)
                Throw New Exceptions.ColorException(DoTranslation("One or more of the colors is invalid.") + " {0}", ex, ex.Message)
            End Try
        Else
            KernelEventManager.RaiseColorSetError(ColorSetErrorReasons.NoColors)
            Throw New InvalidOperationException(DoTranslation("Colors are not available. Turn on colored shell in the kernel config."))
        End If
        Return False
    End Function

End Module

File: public/Nitrocid/ConsoleBase/Themes/ThemeTools.cs
//
// Nitrocid KS  Copyright (C) 2018-2025  Aptivi
//
// This file is part of Nitrocid KS
//
// Nitrocid KS is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Nitrocid KS is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Terminaux.Colors;
using Newtonsoft.Json.Linq;
using Nitrocid.Kernel.Configuration;
using Nitrocid.Kernel.Debugging;
using Nitrocid.Files;
using Nitrocid.Languages;
using Nitrocid.Kernel.Exceptions;
using Nitrocid.ConsoleBase.Colors;
using Nitrocid.Kernel.Events;
using Nitrocid.Misc.Reflection.Internal;

namespace Nitrocid.ConsoleBase.Themes
{
    /// <summary>
    /// Theme tools module
    /// </summary>
    public static class ThemeTools
    {

        internal readonly static Dictionary<string, ThemeInfo> themes = new()
        {
            { "Default", new ThemeInfo(JToken.Parse(ResourcesManager.GetData("Default.json", ResourcesType.Themes) ?? "[]")) },
            { "Dynamic", new ThemeInfo(JToken.Parse(ResourcesManager.GetData("Dynamic.json", ResourcesType.Themes) ?? "[]")) },
            { "NitricAcid", new ThemeInfo(JToken.Parse(ResourcesManager.GetData("NitricAcid.json", ResourcesType.Themes) ?? "[]")) },
        };

        /// <summary>
        /// Gets the installed themes
        /// </summary>
        /// <returns>List of installed themes and their <see cref="ThemeInfo"/> instances</returns>
        public static Dictionary<string, ThemeInfo> GetInstalledThemes() =>
            new(themes);

        /// <summary>
        /// Gets the installed themes by category
        /// </summary>
        /// <param name="category">Category to look for themes</param>
        /// <returns>List of installed themes and their <see cref="ThemeInfo"/> instances</returns>
        public static Dictionary<string, ThemeInfo> GetInstalledThemesByCategory(ThemeCategory category) =>
            themes
                .Where((kvp) => kvp.Value.Category == category)
                .ToDictionary((kvp) => kvp.Key, (kvp) => kvp.Value);

        /// <summary>
        /// Gets the theme information
        /// </summary>
        /// <param name="theme">Theme name</param>
        /// <param name="throwNotFound">Throws an exception if the theme is not found</param>
        public static ThemeInfo GetThemeInfo(string theme, bool throwNotFound = false)
        {
            var themes = GetInstalledThemes();
            if (themes.TryGetValue(theme, out ThemeInfo? resultingTheme))
                return resultingTheme;
            if (throwNotFound)
                throw new KernelException(KernelExceptionType.NoSuchTheme, Translate.DoTranslation("Invalid color template {0}"), theme);
            return themes["Default"];
        }

        /// <summary>
        /// Checks to see if a built-in (or an addon) theme is found
        /// </summary>
        /// <param name="theme">Theme name</param>
        public static bool IsThemeFound(string theme)
        {
            var themes = GetInstalledThemes();
            return themes.ContainsKey(theme);
        }

        /// <summary>
        /// Gets the colors from the theme
        /// </summary>
        /// <param name="theme">Theme name</param>
        public static Dictionary<KernelColorType, Color> GetColorsFromTheme(string theme) =>
            GetColorsFromTheme(GetThemeInfo(theme));

        /// <summary>
        /// Gets the colors from the theme
        /// </summary>
        /// <param name="themeInfo">Theme instance</param>
        public static Dictionary<KernelColorType, Color> GetColorsFromTheme(ThemeInfo themeInfo)
        {
            if (themeInfo.UseAccentTypes.Length > 0)
                themeInfo.UpdateColors();
            return themeInfo.ThemeColors;
        }

        /// <summary>
        /// Sets system colors according to the theme information instance
        /// </summary>
        /// <param name="themeInfo">A specified theme information instance</param>
        /// <param name="dry">Whether to dryly apply the theme or not</param>
        public static void ApplyTheme(ThemeInfo themeInfo, bool dry = false)
        {
            // Check if the console supports true color
            if (Config.MainConfig.ConsoleSupportsTrueColor && themeInfo.TrueColorRequired || !themeInfo.TrueColorRequired)
            {
                // Check to see if the event is finished
                if (themeInfo.IsExpired)
                {
                    DebugWriter.WriteDebug(DebugLevel.E, "Setting event theme in a day that the event finished...");
                    EventsManager.FireEvent(EventType.ThemeSetError, themeInfo.Name, ThemeSetErrorReasons.EventFinished);
                    throw new KernelException(KernelExceptionType.ThemeManagement, Translate.DoTranslation("The theme {0} celebrates an event, but you're either too early or too late to attend. Each year, this theme is accessible from {1}/{2} to {3}/{4}."), themeInfo.Name, themeInfo.StartMonth, themeInfo.StartDay, themeInfo.EndMonth, themeInfo.EndDay);
                }

                // Set colors as appropriate
                DebugWriter.WriteDebug(DebugLevel.I, "Setting colors as appropriate...");
                SetColorsTheme(themeInfo, dry);
            }
            else
            {
                // We're trying to apply true color on unsupported console
                DebugWriter.WriteDebug(DebugLevel.E, "Unsupported console or the terminal doesn't support true color.");
                EventsManager.FireEvent(EventType.ThemeSetError, themeInfo.Name, ThemeSetErrorReasons.ConsoleUnsupported);
                throw new KernelException(KernelExceptionType.UnsupportedConsole, Translate.DoTranslation("The theme {0} needs true color support, but your console doesn't support it."), themeInfo.Name);
            }

            // Raise event
            EventsManager.FireEvent(EventType.ThemeSet, themeInfo.Name);
        }

        /// <summary>
        /// Sets system colors according to the programmed templates
        /// </summary>
        /// <param name="theme">A specified theme</param>
        /// <param name="dry">Whether to dryly apply the theme or not</param>
        public static void ApplyThemeFromResources(string theme, bool dry = false)
        {
            DebugWriter.WriteDebug(DebugLevel.I, "Theme: {0}", theme);
            if (GetInstalledThemes().ContainsKey(theme))
            {
                DebugWriter.WriteDebug(DebugLevel.I, "Theme found.");

                // Populate theme info and use it
                var ThemeInfo = GetThemeInfo(theme);
                ApplyTheme(ThemeInfo, dry);
            }
            else
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Theme not found.");
                EventsManager.FireEvent(EventType.ThemeSetError, theme, ThemeSetErrorReasons.NotFound);
                throw new KernelException(KernelExceptionType.NoSuchTheme, Translate.DoTranslation("Invalid color template {0}"), theme);
            }
        }

        /// <summary>
        /// Sets system colors according to the template file
        /// </summary>
        /// <param name="ThemeFile">Theme file</param>
        /// <param name="dry">Whether to dryly apply the theme or not</param>
        public static void ApplyThemeFromFile(string ThemeFile, bool dry = false)
        {
            try
            {
                DebugWriter.WriteDebug(DebugLevel.I, "Theme file name: {0}", ThemeFile);
                ThemeFile = FilesystemTools.NeutralizePath(ThemeFile, true);
                DebugWriter.WriteDebug(DebugLevel.I, "Theme file path: {0}", ThemeFile);

                // Populate theme info and use it
                var ThemeInfo = new ThemeInfo(ThemeFile);
                ApplyTheme(ThemeInfo, dry);
            }
            catch (FileNotFoundException)
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Theme not found.");
                EventsManager.FireEvent(EventType.ThemeSetError, ThemeFile, ThemeSetErrorReasons.NotFound);
                throw new KernelException(KernelExceptionType.NoSuchTheme, Translate.DoTranslation("Invalid color template {0}"), ThemeFile);
            }
        }

        /// <summary>
        /// Sets custom colors. It only works if colored shell is enabled.
        /// </summary>
        /// <param name="ThemeInfo">Theme information</param>
        /// <param name="dry">Whether to dryly set the colors or not</param>
        /// <exception cref="InvalidOperationException"></exception>
        public static void SetColorsTheme(ThemeInfo ThemeInfo, bool dry = false)
        {
            if (ThemeInfo is null)
                throw new KernelException(KernelExceptionType.Color, nameof(ThemeInfo));

            // Check to see if we're trying to preview theme on non-true color console
            if (MinimumTypeRequired(ThemeInfo, ColorType.TrueColor) && !Config.MainConfig.ConsoleSupportsTrueColor)
                throw new KernelException(KernelExceptionType.UnsupportedConsole, Translate.DoTranslation("Your console must support true color to use this theme."));

            // Set the colors
            try
            {
                for (int typeIndex = 0; typeIndex < Enum.GetValues(typeof(KernelColorType)).Length; typeIndex++)
                {
                    KernelColorType type = KernelColorTools.KernelColors.Keys.ElementAt(typeIndex);
                    var themeColor = ThemeInfo.ThemeColors[type];
                    DebugWriter.WriteDebug(DebugLevel.I, "Kernel color type {0}, setting theme color {1}...", type.ToString(), themeColor.PlainSequence);
                    KernelColorTools.KernelColors[type] = themeColor;
                }
                ColorTools.LoadBack(KernelColorTools.GetColor(KernelColorType.Background));
                if (!dry)
                    Config.CreateConfig();

                // Raise event
                EventsManager.FireEvent(EventType.ColorSet);
            }
            catch (Exception ex)
            {
                DebugWriter.WriteDebugStackTrace(ex);
                EventsManager.FireEvent(EventType.ColorSetError, KernelColorSetErrorReasons.InvalidColors);
                throw new KernelException(KernelExceptionType.Color, Translate.DoTranslation("One or more of the colors is invalid.") + " {0}", ex, ex.Message);
            }
        }

        /// <summary>
        /// Sets custom colors. It only works if colored shell is enabled.
        /// </summary>
        /// <param name="ThemeInfo">Theme information</param>
        /// <returns>True if successful; False if unsuccessful</returns>
        /// <exception cref="InvalidOperationException"></exception>
        public static bool TrySetColorsTheme(ThemeInfo ThemeInfo)
        {
            try
            {
                SetColorsTheme(ThemeInfo);
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }

        /// <summary>
        /// Is the 255 color support required?
        /// </summary>
        /// <param name="theme">Theme instance</param>
        /// <param name="type">Minimum required color type</param>
        /// <returns>If required, then true. Always false when <see cref="ColorType.FourBitColor"/> is passed.</returns>
        public static bool MinimumTypeRequired(string theme, ColorType type) =>
            MinimumTypeRequired(GetThemeInfo(theme), type);

        /// <summary>
        /// Is the 255 color support required?
        /// </summary>
        /// <param name="theme">Theme instance</param>
        /// <param name="type">Minimum required color type</param>
        /// <returns>If required, then true. Always false when <see cref="ColorType.FourBitColor"/> is passed.</returns>
        public static bool MinimumTypeRequired(ThemeInfo theme, ColorType type) =>
            MinimumTypeRequired(GetColorsFromTheme(theme), type);

        /// <summary>
        /// Is the 255 color support required?
        /// </summary>
        /// <param name="colors">Dictionary of colors</param>
        /// <param name="type">Minimum required color type</param>
        /// <returns>If required, then true. Always false when <see cref="ColorType.FourBitColor"/> is passed.</returns>
        public static bool MinimumTypeRequired(Dictionary<KernelColorType, Color> colors, ColorType type)
        {
            if (type == ColorType.FourBitColor)
                return false;

            // Now, check for 255-color requirement
            for (int key = 0; key < colors.Count; key++)
                if (colors.Values.ElementAt(key).Type <= type)
                    return true;

            // Else, 255 color support is not required
            return false;
        }

    }
}
